src/mms_client/utils/auditing.py

```diff
@@ -0,0 +1,54 @@
+"""Contains functions for auditability plugins."""
+
+from abc import ABC
+from abc import abstractmethod
+
+from lxml.etree import _Element as Element
+from lxml.etree import tostring
+from zeep import Plugin
+
+
+class AuditPlugin(ABC, Plugin):
+    """Base class for audit plugins."""
+
+    def egress(self, envelope: Element, http_headers: dict, operation, binding_options):
+        """Handle the MMS request before it is sent.
+
+        Arguments are the same as in the egress method of the Plugin class.
+
+        Returns:
+        lxml.etree.Element: The XML message to send.
+        dict:               The HTTP headers to send.
+        """
+        data = tostring(envelope, encoding="UTF-8", xml_declaration=True)
+        self.audit_request(data)
+        return envelope, http_headers
+
+    def ingress(self, envelope: Element, http_headers: dict, operation):
+        """Handle the MMS response before it is processed.
+
+        Arguments are the same as in the ingress method of the Plugin class.
+
+        Returns:
+        lxml.etree.Element: The XML message to process.
+        dict:               The HTTP headers to process.
+        """
+        data = tostring(envelope, encoding="UTF-8", xml_declaration=True)
+        self.audit_response(data)
+        return envelope, http_headers
+
+    @abstractmethod
+    def audit_request(self, mms_request: bytes) -> None:
```

Comment:
What is the intended use for these abstract methods?

Reply:
Essentially, what it means is that the user creates an object that inherits from `AuditPlugin` and writes their own implementation of `audit_request` and `audit_response` that do whatever they want with the request and response objects, without necessarily having to re-implement `ingress` and `egress`. See the attached test for more info.